src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+
+        public void addRecords ( in void[] record_batch )
+        {
+            // Append record_batch to *this.batch_buffer, which may or may not
+            // be empty.
+
+            if (!(*this.batch_buffer).length)
+                enableStomping(*this.batch_buffer);
+
+            // Append record_batch, then set this.remaining_batch to reference
+            // the remaining records. To avoid a dangling slice if
+            // *this.batch_buffer is relocated, set this.remaining_batch to null
+            // first.
+            size_t n_processed = (*this.batch_buffer).length - this.remaining_batch.length;
+            this.remaining_batch = null;
+            (*this.batch_buffer) ~= record_batch;
+            this.remaining_batch = (*this.batch_buffer)[n_processed .. $];
+
+            if (this.fiber_suspended == fiber_suspended.WaitingForRecords)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Resumes passing records to the user. Called when the user resumes
+            the request through the controller.
+
+        ***********************************************************************/
+
+        public void resume ( )
+        {
+            if (this.fiber_suspended == fiber_suspended.RequestSuspended)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Requests the fiber to terminate when all remaining records have been
+            passed to the user. Called when a `Stopped` message from the node
+            has arrived.
+
+        ***********************************************************************/
+
+        public void stop ( )
+        {
+            this.stopped = true;
+            if (this.fiber_suspended == fiber_suspended.WaitingForRecords)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Waits for a batch of records to arrive and passes it to the user,
+            then sends the `Continue` message to the node, in a loop. Handles
+            suspending the request through the controller, for resuming call
+            `resume`. Calling `stop` makes this routine terminate after all
+            remaining records have been passed to the user.
+
+        ***********************************************************************/
+
+        private void fiberMethod ( )
+        {
+            while (this.waitForRecords())
+            {
+                for (uint yield_count = 1; ; yield_count++)
+                {
+                    if (this.outer.context.shared_working.suspendable_control.suspended)
+                    {
+                        yield_count = 0;
+                        this.fiber_suspended = fiber_suspended.RequestSuspended;
```

Comment:
What happens if the controller is used to stop the request while it's suspended at this point?

Reply:
When the controller is used to stop the request, the only thing that happens is that `Controller` sends a `Stop` message to the node. You probably mean, what happens if a `Stopped` message is received at this point: `this.stopped` is set to `true`, nothing else; the fiber stays suspended until the controller is used to resume the request.
This loop ignores the "stopped" condition. It simply passes all remaining records in the batch to the user, allowing suspending and resuming using the controller. When all remaining records in the batch have been processed and this loop exits, the outer loop notices `this.stopped == true` and exits, then the fiber terminates.

Reply:
Ah yes, indeed, I was referring to what happens if `stop` is called at this point.

Yes, I'd forgotten that it's valid to suspend/resume _after_ stopping. _Maybe_ this should be mentioned somewhere? Not sure where though...